### PR TITLE
Add support for nested logic and group errors

### LINKED ISSUE #278: Seeing "TypeError: exceptions must derive from BaseException" when looking at validator.errors on failed validation
This issue tracker is dedicated to bugs and feature requests. 

Stack Overflow is the go-to place for general questions and general how-tos. On
SO, questions tagged with the 'cerberus' tag are actively monitored by the
project author, contributors and users.

When reporting a bug, please post an example and make it as simple as possible.
From your use-case throw everything out that is irrelevant for your question
/ isn't necessary to evoke the bug.

Try to adhere as much as possible to the template below:

Used Cerberus version / latest commit: ...
- I consulted these documentations:
  - [x] http://docs.python-cerberus.org/en/stable/
  - [x] http://docs.python-cerberus.org/en/latest/
- I consulted these sections of the docs (add more lines as necessary):
  - ...
  - ...
- [x] I found nothing relevant to my problem in the docs.
- [x] I found the documentation not helpful to my problem.
- [x] I have the capacity to improve the docs when my problem is solved.
- [x] I have the capacity to submit a patch when a bug is identified.

---
#### Use-case abstract

I am using a 'oneof_schema' rule in order to allow either:
- an empty dictionary, or
- a dictionary of a specific format

---
#### Support request / Bug report

Using: Cerberus 1.0.1 and Python 3.5.2

I'm seeing a strange error when trying to use a 'oneof_schema' rule for my document. The goal is to either allow the 'measurement' field to be an empty dict, or following a specific format.

Examples of documents that (should) pass:

```
{
    'tokens': [
        {
            'measurement': {
                'qty': { 'num': 10, 'den': 1, },
                'units': 'lbs',
            },
        },
    ],
}

{
    'tokens': [
        {
            'measurement': {
                'qty': 100,
                'units': 'lbs',
            },
        },
    ],
}

{
    'tokens': [
        {
            'measurement': {},
        },
    ],
}
```

Example of a document that should fail:

```
{
    'tokens': [
        {
            'measurement': {
                'qty':     { 'num': 'THIS WILL FAIL!', 'den': 1, },
                'units':    'lbs',
            },
        },
    ],
}
```

Interestingly, when I tested the above "should-fail" document, I get a strange exception:

```
$ python failure.py
Traceback (most recent call last):
  File "main.py", line 64, in <module>
    print(v.errors)
  File "/home/cheffu/venvs/cerberus-cheffu-test/lib/python3.5/site-packages/cerberus/validator.py", line 360, in errors
    return self.error_handler(self._errors)
  File "/home/cheffu/venvs/cerberus-cheffu-test/lib/python3.5/site-packages/cerberus/errors.py", line 443, in __call__
    self.extend(errors)
  File "/home/cheffu/venvs/cerberus-cheffu-test/lib/python3.5/site-packages/cerberus/errors.py", line 372, in extend
    self.add(error)
  File "/home/cheffu/venvs/cerberus-cheffu-test/lib/python3.5/site-packages/cerberus/errors.py", line 453, in add
    self.insert_group_error(error)
  File "/home/cheffu/venvs/cerberus-cheffu-test/lib/python3.5/site-packages/cerberus/errors.py", line 496, in insert_group_error
    self.insert_logic_error(error)
  File "/home/cheffu/venvs/cerberus-cheffu-test/lib/python3.5/site-packages/cerberus/errors.py", line 516, in insert_logic_error
    raise NotImplemented
TypeError: exceptions must derive from BaseException
```

If I change the rule to remove the 'oneof_schema' rule (losing the ability to accept an empty dict), I'm able to see the validation error, as expected:

```
$ python ok.py
{'tokens': [{0: [{'measurement': [{'qty': [{'oneof': ['none or more than one rule validate', {'oneof definition 1': ['must be of integer type'], 'oneof definition 0': ['must be of integer type']}]}]}]}]}]}
```

The only difference in the code between the two files is the 'oneof_schema' rule in the 'measurement' schema (lines 36-39 in failure.py). Is this an error on my part, or something more insiduous?

Attached are the two Python files: [failure.py.txt](https://github.com/nicolaiarocci/cerberus/files/557555/failure.py.txt) and [ok.py.txt](https://github.com/nicolaiarocci/cerberus/files/557556/ok.py.txt).

### PR DESCRIPTION
**Pull request by kornholi**
Should fix #278 and #299. Not sure how I feel about modifying `document_path`, but I don't have the time right now to come up with something better.

### CODE REVIEW

File: cerberus/tests/test_validation.py [hunk 1]
```diff
@@ -1276,6 +1276,39 @@ def test_nested_oneof_type():
...
+                'schema': {'bar': {'oneof_type': ['integer', 'float']}}
+            }},
+            {'baz': {'type': 'string'}}
+        ]}}
+
+    document = {'abc': {'foo': {'bar': 'bad'}}}
+
+    expected_errors = {
+        'abc': [
+            {'oneof': [
```
Comment (funkyfuture) [inline cerberus/tests/test_validation.py]: do we really need this level?

File: cerberus/tests/test_validation.py [hunk 2]
```diff
@@ -1276,6 +1276,39 @@ def test_nested_oneof_type():
...
+                'none or more than one rule validate',
+                {'oneof definition 0': [
+                    {'foo': [{'bar': [
+                        {'oneof': [
+                            'none or more than one rule validate',
```
Comment (funkyfuture) [inline cerberus/tests/test_validation.py]: not your fault, but it should be 'validated'.

File: cerberus/errors.py [hunk 3]
```diff
@@ -532,25 +534,58 @@ def insert_group_error(self, error):
...
+
+    def rewrite_logic_error_path(self, error, offset=0):
+        error.document_path = error.document_path + (error.rule, )
+        child_start = len(error.document_path) - offset - 1
+
+        for i, child_errors in error.definitions_errors.items():
+            if not child_errors:
+                continue
+
+            nodename = '%s definition %s' % (error.rule, i)
```
Comment (funkyfuture) [inline cerberus/errors.py]: i'm not sure whether there is a stringent way to format strings in the code base. if there's only or mostly `.format` used so far, please do the same.

**Comment by funkyfuture** [comment]
i'm not getting the logs of the failed test against CPython 3.3. in doubt, i'm in favour of dropping its support.

File: cerberus/tests/test_validation.py (hunk 2 above)
Comment (kornholi) [inline cerberus/tests/test_validation.py]: all of the other errors are in present tense

File: cerberus/tests/test_validation.py (hunk 1 above)
Comment (kornholi) [inline cerberus/tests/test_validation.py]: line length

File: cerberus/errors.py (hunk 3 above)
Comment (kornholi) [inline cerberus/errors.py]: from a quick `rg` it looks like an even mix

**Comment by kornholi** [comment]
Not sure what you mean about changing bare errors. They should stay exactly the same, and none of the tests changed. Do you have an example?

I've been testing this with pretty complicated nested schemas and haven't hit any edge cases yet.

File: cerberus/tests/test_validation.py (hunk 1 above)
Comment (funkyfuture) [inline cerberus/tests/test_validation.py]: ah, i didn't mean indentation level in the code, but the level in the tree representation. the following two lines also mention the oneof rule. this isn't a request to remove it, but i want to raise the question.

**Comment by funkyfuture** [comment]
interesting how fast the memory about one own's writing fades.. thanks for your checks on my misleading remarks.

when i'm not wrong:

- the handler iterates over the 'bare' error representations from `Validator._errors`
- these are used as sources to select and format the descriptional strings of the handler
- but in lines 568 and 573 you change the `document_path` attribute of the original error representation
- thus the original information is altered with a representational variation that is specific to the handler

the simplest solution is propably to add this method to `errors.ValidationError`:

```python
def __copy__
    return self.__class__(**vars(self))
```

and then use `copy.copy` to get an instance that can be manipulated. though this proposal isn't correct. the attributes `constraint`, `value` and `info` should be deepcopied before getting passed to the initialization to avoid problems with mutable values.

> I've been testing this with pretty complicated nested schemas and haven't hit any edge cases yet.

perfect. if these are sensible for you application, feel free to add reflecting regression tests.

File: cerberus/tests/test_validation.py (hunk 1 above)
Comment (kornholi) [inline cerberus/tests/test_validation.py]: yeah I was thinking the same thing. maybe change the children to just `definition n`?

Comment (funkyfuture) [inline cerberus/tests/test_validation.py]: but then different of-rule-related errors (eg `oneof` and `anyof`) couldn't be discriminated.

Comment (kornholi) [inline cerberus/tests/test_validation.py]: Right, I'm saying keep the `oneof` level but change `oneof definition n` to `definition n`.

**Comment by kornholi** [comment]
Would [this deepcopy](https://github.com/pyeve/cerberus/pull/321/files#diff-b21a18a810079ce89a6e2d289ba98e93R481) be enough or am I missing something?

File: cerberus/tests/test_validation.py (hunk 1 above)
Comment (funkyfuture) [inline cerberus/tests/test_validation.py]: which will not work (and not communicate well if somehow mangled) if there are e.g. a `oneof` and an `anyof` rule for the same field and both yield errors.

**Comment by funkyfuture** [comment]
i'm surprised, but it works:

```python
>>> from cerberus.errors import ValidationError
>>> from copy import deepcopy
>>> e = ValidationError((), (), 0, 'foo', {'x': ['a', 'b', 'c']}, None, ())
>>> f = deepcopy(e)
>>> f.constraint['x'].append('d')
>>> 'd' in e.constraint['x']
False
```

**Comment by kornholi** [comment]
Is there anything left to do here? Not sure what's going on with the 3.5 tests.

File: cerberus/tests/test_validation.py (hunk 1 above)
Comment (funkyfuture) [inline cerberus/tests/test_validation.py]: this is still open, right?

**Comment by funkyfuture** [comment]
> Not sure what's going on with the 3.5 tests.

this is #281

**Comment by nicolaiarocci** [comment]
@kornholi thanks for your contribution!
@funkyfuture is this ready for merge? Let's ignore the failing doctest for the time being.

**Comment by funkyfuture** [comment]
the redundant information and extra tree level of *of-errors is still [unsolved](https://github.com/pyeve/cerberus/pull/321#discussion_r124809777).

**Comment by kornholi** [comment]
Could you provide a concrete example of how you want it to look? Unless I am misunderstanding you, it seems out of scope for this PR.

**Comment by funkyfuture** [comment]
```python
expected_errors = {
    'abc': [
            ['none or more than one rule validate',
             {'oneof definition 0':
              [{'foo':
                [{'bar':
                 [{'oneof':
                  ['none or more than one rule validate',
                   {'oneof definition 0': ['must be of integer type'],
                    'oneof definition 1': ['must be of float type']}
                   ]}]}]}],
              'oneof definition 1': [{'foo': ['unknown field']}]}]
    ]
}
```

**Comment by funkyfuture** [comment]
i pledge to mark this for the `1.2` milestone.

@kornholi maybe a refactoring of `Validator._drop_nodes_from_errorpaths`'s  mechanic would help with the superfluous docuemnt level.

**Comment by nicolaiarocci** [comment]
Could you please rebase on top of master? Also it is my understanding that there's still some refactoring needed, see @funkyfuture comment above.

**Comment by nicolaiarocci** [comment]
Rebased (once again) and merged. Thanks!
